BLD: PPC issues with new CPU dispatcher

### Describe the issue:

Collecting here multiple issues I observed that prevent building on PPC with the new CPU dispatcher.

Issue 1:

Building on a POWER9 host which supports VSX3, VSX4 is incorrectly detected as supported.

```
Host machine cpu family: ppc64
Host machine cpu: ppc64le
Program python found: YES (/usr/bin/python3.11)
Found pkg-config: /usr/bin/pkg-config (2.0.3)
Run-time dependency python found: YES 3.11
Has header "Python.h" with dependency python-3.11: YES
Compiler for C supports arguments -fno-strict-aliasing: YES
Message: Appending option "detect" to "cpu-baseline" due to detecting global architecture c_arg "-mcpu=native"
Test features "VSX VSX2 VSX3 VSX4" : Supported
Message: During parsing cpu-dispatch: The following CPU features were ignored due to platform incompatibility or lack of support:
"XOP FMA4"
Test features "VSX VSX2 VSX3 VSX4" : Supported
Configuring npy_cpu_dispatch_config.h using configuration
Message:
CPU Optimization Options
  baseline:
    Requested : min+detect
    Enabled   : VSX VSX2 VSX3 VSX4
  dispatch:
    Requested : max -xop -fma4
    Enabled   :
```

Issue 2:

Even in VSX3-conditional code, build fails with an incorrect variable name in ASM code.

```
[9/297] powerpc64le-unknown-linux-gnu-g++ -Inumpy/core/libnpymath.a.p -Inumpy/core -I../numpy-1.26.0/numpy/core -Inumpy/core/include -I../numpy-1.26.0/numpy/core/include -I../numpy-1.26.0/nump
y/core/src/npymath -I../numpy-1.26.0/numpy/core/src/common -I/usr/include/python3.11 -I/var/tmp/portage/dev-python/numpy-1.26.0/work/numpy-1.26.0-python3_11/meson_cpu -fdiagnostics-color=alway
s -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -std=c++17 -O3 -mcpu=power10 -DNPY_HAVE_VSX2 -DNPY_HAVE_VSX -DNPY_HAVE_VSX_ASM -DNPY_HAVE_VSX3 -DNPY_HAVE_VSX4 -DNPY_HAVE_VSX4_MMA -O3 -mcpu=native
 -mtune=native -pipe -fno-strict-aliasing -DNDEBUG -fPIC -MD -MQ numpy/core/libnpymath.a.p/src_npymath_halffloat.cpp.o -MF numpy/core/libnpymath.a.p/src_npymath_halffloat.cpp.o.d -o numpy/core
/libnpymath.a.p/src_npymath_halffloat.cpp.o -c ../numpy-1.26.0/numpy/core/src/npymath/halffloat.cpp
FAILED: numpy/core/libnpymath.a.p/src_npymath_halffloat.cpp.o 
powerpc64le-unknown-linux-gnu-g++ -Inumpy/core/libnpymath.a.p -Inumpy/core -I../numpy-1.26.0/numpy/core -Inumpy/core/include -I../numpy-1.26.0/numpy/core/include -I../numpy-1.26.0/numpy/core/s
rc/npymath -I../numpy-1.26.0/numpy/core/src/common -I/usr/include/python3.11 -I/var/tmp/portage/dev-python/numpy-1.26.0/work/numpy-1.26.0-python3_11/meson_cpu -fdiagnostics-color=always -D_FIL
E_OFFSET_BITS=64 -Wall -Winvalid-pch -std=c++17 -O3 -mcpu=power10 -DNPY_HAVE_VSX2 -DNPY_HAVE_VSX -DNPY_HAVE_VSX_ASM -DNPY_HAVE_VSX3 -DNPY_HAVE_VSX4 -DNPY_HAVE_VSX4_MMA -O3 -mcpu=native -mtune=
native -pipe -fno-strict-aliasing -DNDEBUG -fPIC -MD -MQ numpy/core/libnpymath.a.p/src_npymath_halffloat.cpp.o -MF numpy/core/libnpymath.a.p/src_npymath_halffloat.cpp.o.d -o numpy/core/libnpym
ath.a.p/src_npymath_halffloat.cpp.o -c ../numpy-1.26.0/numpy/core/src/npymath/halffloat.cpp
In file included from ../numpy-1.26.0/numpy/core/src/common/common.hpp:10,
                 from ../numpy-1.26.0/numpy/core/src/npymath/halffloat.cpp:12:
../numpy-1.26.0/numpy/core/src/common/half.hpp: In member function ‘np::Half::operator double() const’:
../numpy-1.26.0/numpy/core/src/common/half.hpp:116:38: error: ‘vf32’ was not declared in this scope
  116 |                              : "=wa"(vf32)
      |                                      ^~~~
ninja: build stopped: subcommand failed.
```

Issue 3:

After applying the following patch to fix:

```
diff --git a/numpy/core/src/common/half.hpp b/numpy/core/src/common/half.hpp
index 054c99859..b85b592bb 100644
--- a/numpy/core/src/common/half.hpp
+++ b/numpy/core/src/common/half.hpp
@@ -113,7 +113,7 @@ class Half final {
     #elif defined(NPY_HAVE_VSX3) && defined(NPY_HAVE_VSX_ASM)
         __vector float vf64;
         __asm__ __volatile__("xvcvhpdp %x0,%x1"
-                             : "=wa"(vf32)
+                             : "=wa"(vf64)
                              : "wa"(vec_splats(bits_)));
         return vec_extract(vf64, 0);
     #else
```

it would seem that the assembly uses invalid instructions.

```
[9/297] powerpc64le-unknown-linux-gnu-g++ -Inumpy/core/libnpymath.a.p -Inumpy/core -I../numpy-1.26.0/numpy/core -Inumpy/core/include -I../numpy-1.26.0/numpy/core/include -I../numpy-1.26.0/nump
y/core/src/npymath -I../numpy-1.26.0/numpy/core/src/common -I/usr/include/python3.11 -I/var/tmp/portage/dev-python/numpy-1.26.0/work/numpy-1.26.0-python3_11/meson_cpu -fdiagnostics-color=alway
s -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -std=c++17 -O3 -mcpu=power10 -DNPY_HAVE_VSX2 -DNPY_HAVE_VSX -DNPY_HAVE_VSX_ASM -DNPY_HAVE_VSX3 -DNPY_HAVE_VSX4 -DNPY_HAVE_VSX4_MMA -O3 -mcpu=native
 -mtune=native -pipe -fno-strict-aliasing -DNDEBUG -fPIC -MD -MQ numpy/core/libnpymath.a.p/src_npymath_halffloat.cpp.o -MF numpy/core/libnpymath.a.p/src_npymath_halffloat.cpp.o.d -o numpy/core
/libnpymath.a.p/src_npymath_halffloat.cpp.o -c ../numpy-1.26.0/numpy/core/src/npymath/halffloat.cpp
FAILED: numpy/core/libnpymath.a.p/src_npymath_halffloat.cpp.o 
powerpc64le-unknown-linux-gnu-g++ -Inumpy/core/libnpymath.a.p -Inumpy/core -I../numpy-1.26.0/numpy/core -Inumpy/core/include -I../numpy-1.26.0/numpy/core/include -I../numpy-1.26.0/numpy/core/s
rc/npymath -I../numpy-1.26.0/numpy/core/src/common -I/usr/include/python3.11 -I/var/tmp/portage/dev-python/numpy-1.26.0/work/numpy-1.26.0-python3_11/meson_cpu -fdiagnostics-color=always -D_FIL
E_OFFSET_BITS=64 -Wall -Winvalid-pch -std=c++17 -O3 -mcpu=power10 -DNPY_HAVE_VSX2 -DNPY_HAVE_VSX -DNPY_HAVE_VSX_ASM -DNPY_HAVE_VSX3 -DNPY_HAVE_VSX4 -DNPY_HAVE_VSX4_MMA -O3 -mcpu=native -mtune=
native -pipe -fno-strict-aliasing -DNDEBUG -fPIC -MD -MQ numpy/core/libnpymath.a.p/src_npymath_halffloat.cpp.o -MF numpy/core/libnpymath.a.p/src_npymath_halffloat.cpp.o.d -o numpy/core/libnpym
ath.a.p/src_npymath_halffloat.cpp.o -c ../numpy-1.26.0/numpy/core/src/npymath/halffloat.cpp
{standard input}: Assembler messages:
{standard input}:38: Error: unrecognized opcode: `xvcvhpdp'
{standard input}:83: Error: unrecognized opcode: `xvcvdphp'
{standard input}:846: Error: unrecognized opcode: `xvcvdphp'
{standard input}:892: Error: unrecognized opcode: `xvcvhpdp'
ninja: build stopped: subcommand failed.
```

This is using the following binutils:

```
$ ld --version
GNU ld (Gentoo 2.41 p2) 2.41.0
Copyright (C) 2023 Free Software Foundation, Inc.
This program is free software; you may redistribute it under the terms of
the GNU General Public License version 3 or (at your option) a later version.
This program has absolutely no warranty.
```

All three of these issues apply on both big and little endian, as well as regardless of using `-mcpu=native` or `-mcpu=power9`.  As always, I provide [free shell access to the hardware I encountered these issues on](https://static.matoro.tk/isa-sandbox-faq.html) if desired.

@seiko2plus @rgommers 

### Reproduce the code example:

```python
N/A
```


### Error message:

_No response_

### Runtime information:

N/A

### Context for the issue:

_No response_

--- a/.github/workflows/linux_qemu.yml
+++ b/.github/workflows/linux_qemu.yml
@@ -52,6 +52,13 @@ jobs:
               "test_kind or test_multiarray or test_simd or test_umath or test_ufunc",
           ]
           - [
+              "ppc64le - baseline(Power9)",
+              "powerpc64le-linux-gnu",
+              "ppc64le/ubuntu:22.04",
+              "-Dallow-noblas=true -Dcpu-baseline=vsx3",
+              "test_kind or test_multiarray or test_simd or test_umath or test_ufunc",
+          ]
+          - [
               "s390x",
               "s390x-linux-gnu",
               "s390x/ubuntu:22.04",

--- a/meson_cpu/ppc64/meson.build
+++ b/meson_cpu/ppc64/meson.build
@@ -26,6 +26,9 @@ VSX3 = mod_features.new(
   'VSX3', 3, implies: VSX2, args: {'val': '-mcpu=power9', 'match': '.*[mcpu=|vsx].*'},
   detect: {'val': 'VSX3', 'match': 'VSX.*'},
   test_code: files(source_root + '/numpy/distutils/checks/cpu_vsx3.c')[0],
+  extra_tests: {
+    'VSX3_HALF_DOUBLE': files(source_root + '/numpy/distutils/checks/extra_vsx3_half_double.c')[0]
+  }
 )
 VSX4 = mod_features.new(
   'VSX4', 4, implies: VSX3, args: {'val': '-mcpu=power10', 'match': '.*[mcpu=|vsx].*'},

--- a/numpy/_core/src/common/half.hpp
+++ b/numpy/_core/src/common/half.hpp
@@ -36,7 +36,7 @@ class Half final {
         #endif
         ) || (
             std::is_same_v<T, double> &&
-        #if defined(NPY_HAVE_AVX512FP16) || defined(NPY_HAVE_VSX3)
+        #if defined(NPY_HAVE_AVX512FP16) || (defined(NPY_HAVE_VSX3) && defined(NPY_HAVE_VSX3_HALF_DOUBLE))
             true
         #else
             false
@@ -73,11 +73,8 @@ class Half final {
     #if defined(NPY_HAVE_AVX512FP16)
         __m128d md = _mm_load_sd(&f);
         bits_ = static_cast<uint16_t>(_mm_cvtsi128_si32(_mm_castph_si128(_mm_cvtpd_ph(md))));
-    #elif defined(NPY_HAVE_VSX3) && defined(NPY_HAVE_VSX_ASM)
-        __vector double vf64 = vec_splats(f);
-        __vector unsigned short vf16;
-        __asm__ __volatile__ ("xvcvdphp %x0,%x1" : "=wa" (vf16) : "wa" (vf64));
-        bits_ = vec_extract(vf16, 0);
+    #elif defined(NPY_HAVE_VSX3) && defined(NPY_HAVE_VSX3_HALF_DOUBLE)
+        __asm__ __volatile__ ("xscvdphp %x0,%x1" : "=wa" (bits_) : "wa" (f));
     #else
         bits_ = half_private::FromDoubleBits(BitCast<uint64_t>(f));
     #endif
@@ -96,7 +93,7 @@ class Half final {
         __vector float vf32;
         __asm__ __volatile__("xvcvhpsp %x0,%x1"
                              : "=wa"(vf32)
-                             : "wa"(vec_splats(bits_.u)));
+                             : "wa"(vec_splats(bits_)));
         return vec_extract(vf32, 0);
     #else
         return BitCast<float>(half_private::ToFloatBits(bits_));
@@ -110,12 +107,12 @@ class Half final {
         double ret;
         _mm_store_sd(&ret, _mm_cvtph_pd(_mm_castsi128_ph(_mm_cvtsi32_si128(bits_))));
         return ret;
-    #elif defined(NPY_HAVE_VSX3) && defined(NPY_HAVE_VSX_ASM)
-        __vector float vf64;
-        __asm__ __volatile__("xvcvhpdp %x0,%x1"
-                             : "=wa"(vf32)
-                             : "wa"(vec_splats(bits_)));
-        return vec_extract(vf64, 0);
+    #elif defined(NPY_HAVE_VSX3) && defined(NPY_HAVE_VSX3_HALF_DOUBLE)
+        double f64;
+        __asm__ __volatile__("xscvhpdp %x0,%x1"
+                             : "=wa"(f64)
+                             : "wa"(bits_));
+        return f64;
     #else
         return BitCast<double>(half_private::ToDoubleBits(bits_));
     #endif

--- a/numpy/_core/tests/test_half.py
+++ b/numpy/_core/tests/test_half.py
@@ -21,8 +21,11 @@ class TestHalf:
         # An array of all possible float16 values
         self.all_f16 = np.arange(0x10000, dtype=uint16)
         self.all_f16.dtype = float16
-        self.all_f32 = np.array(self.all_f16, dtype=float32)
-        self.all_f64 = np.array(self.all_f16, dtype=float64)
+
+        # NaN value can cause an invalid FP exception if HW is been used
+        with np.errstate(invalid='ignore'):
+            self.all_f32 = np.array(self.all_f16, dtype=float32)
+            self.all_f64 = np.array(self.all_f16, dtype=float64)
 
         # An array of all non-NaN float16 values, in sorted order
         self.nonan_f16 = np.concatenate(
@@ -44,14 +47,19 @@ class TestHalf:
         # value is preserved when converting to/from other floats.
 
         # Convert from float32 back to float16
-        b = np.array(self.all_f32, dtype=float16)
-        assert_equal(self.all_f16.view(dtype=uint16),
-                     b.view(dtype=uint16))
+        with np.errstate(invalid='ignore'):
+            b = np.array(self.all_f32, dtype=float16)
+        # avoid testing NaNs due to differ bits wither Q/SNaNs
+        b_nn = b == b
+        assert_equal(self.all_f16[b_nn].view(dtype=uint16),
+                     b[b_nn].view(dtype=uint16))
 
         # Convert from float64 back to float16
-        b = np.array(self.all_f64, dtype=float16)
-        assert_equal(self.all_f16.view(dtype=uint16),
-                     b.view(dtype=uint16))
+        with np.errstate(invalid='ignore'):
+            b = np.array(self.all_f64, dtype=float16)
+        b_nn = b == b
+        assert_equal(self.all_f16[b_nn].view(dtype=uint16),
+                     b[b_nn].view(dtype=uint16))
 
         # Convert float16 to longdouble and back
         # This doesn't necessarily preserve the extra NaN bits,

--- a/numpy/distutils/ccompiler_opt.py
+++ b/numpy/distutils/ccompiler_opt.py
@@ -301,7 +301,8 @@ class _Config:
         ## Power8/ISA 2.07
         VSX2 = dict(interest=2, implies="VSX", implies_detect=False),
         ## Power9/ISA 3.00
-        VSX3 = dict(interest=3, implies="VSX2", implies_detect=False),
+        VSX3 = dict(interest=3, implies="VSX2", implies_detect=False,
+                    extra_checks="VSX3_HALF_DOUBLE"),
         ## Power10/ISA 3.1
         VSX4 = dict(interest=4, implies="VSX3", implies_detect=False,
                     extra_checks="VSX4_MMA"),

--- a/numpy/distutils/checks/extra_vsx3_half_double.c
+++ b/numpy/distutils/checks/extra_vsx3_half_double.c
@@ -1,0 +1,12 @@
+/**
+ * Assembler may not fully support the following VSX3 scalar
+ * instructions, even though compilers report VSX3 support.
+ */
+int main(void)
+{
+    unsigned short bits = 0xFF;
+    double f;
+    __asm__ __volatile__("xscvhpdp %x0,%x1" : "=wa"(f) : "wa"(bits));
+    __asm__ __volatile__ ("xscvdphp %x0,%x1" : "=wa" (bits) : "wa" (f));
+    return bits;
+}

